Source/Tools/LinuxEmulation/LinuxSyscalls/ThreadManager.h

```diff
@@ -22,6 +22,79 @@ namespace FEX::HLE {
 class SyscallHandler;
 class SignalDelegator;
 
+// A latch that can be inspected to see if there is a waiter currently active.
+// This allows us to remove the race condition between a thread trying to go asleep and something else telling it to go to sleep or wake up.
+//
+// Only one thread can ever wait on a latch, while another thread signals it.
+class InspectableLatch final {
+public:
+  bool Wait(struct timespec* Timeout = nullptr) {
+    while (true) {
+      uint32_t Expected = HAS_NO_WAITER;
+      const uint32_t Desired = HAS_WAITER;
+
+      if (Mutex.compare_exchange_strong(Expected, Desired)) {
+        // We have latched, now futex.
+        constexpr int Op = FUTEX_WAIT | FUTEX_PRIVATE_FLAG;
+        // WAIT will keep sleeping on the futex word while it is `val`
+        int Result = ::syscall(SYS_futex, &Mutex, Op,
+                               Desired, // val
+                               Timeout, // Timeout/val2
+                               nullptr, // Addr2
+                               0);      // val3
+
+        if (Timeout && Result == -1 && errno == ETIMEDOUT) {
+          return false;
+        }
+      } else if (Expected == HAS_SIGNALED) {
+        // Reset the latch once signaled
+        Mutex.store(HAS_NO_WAITER);
+        return true;
+      }
```

Comment:
Should there be an assert against `Expected != HAS_NO_WAITER && Expected != HAS_SIGNALED` here to cheaply detect if multiple waiters are active?